segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,179 @@
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * Rules:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int updateParallelism() {
+        // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+        final long time = this.timeSupplier.get();
+        final Snapshot lastSnapshot = this.lastSnapshot.get();
+        final long elapsedMillis = (time - lastSnapshot.timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+        if (elapsedMillis < MIN_FREQUENCY_MILLIS) {
+            // No need to do any change.
+            return lastSnapshot.parallelism;
+        } else if (elapsedMillis >= STALE_MILLIS) {
+            // Last snapshot is too old. We can't make any good determination.
+            resetSnapshot();
+            return this.lastSnapshot.get().parallelism;
+        }
+
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final double throughputDifference = recentThroughput - lastSnapshot.throughput;
+
+        int parallelism = lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
+            // Throughput changed significantly.
+            // But first, check for a special case. If both Throughput and Fill Ratio are decreasing, then most likely the
+            // former is a result of the latter. As such, don't do anything, in that case.
+            boolean fillRatioDecreased = (recentFillRatio - lastSnapshot.fillRatio) <= -SIGNIFICANT_DIFFERENCE;
+            if (throughputDifference > 0 || !fillRatioDecreased) {
+                // Either the throughput increased or the Fill Ratio did not decrease significantly.
+                // Change the parallelism in the same direction as Throughput.
+                // Update the degree of parallelism, but make sure we don't exceed the given bounds.
+                parallelism = MathHelpers.minMax(parallelism + (int) Math.signum(throughputDifference), this.minParallelism, this.maxParallelism);
```

Comment:
I used 2 as an example, you can clearly set the additive increase to 1, in which case it should eventually converge to 7 if that's the optimal value.

Also, I'm not sure why it is going 6 -> 8 -> 4 -> 6 -> 8 -> 4. Specifically, the drop of 8 to 4 is unclear as with a factor of 2, it should be 6 -> 8 -> 6 -> 8 -> 6.

Note that the additive factor doesn't need to be the same as the multiplicative one. The additive factor can be 1 while the multiplicative one can be 2 and vice-versa. The important property is that increases conservatively and drops aggressively. We want the aggressive drop because we don't want to overload our components.